docs/source/daml/intro/3_Data.rst

```diff
@@ -0,0 +1,202 @@
+
+.. literalinclude:: daml/3_Data/Record.daml
+  :language: daml
+  :start-after: -- RECORD_TEST_BEGIN
+  :end-before: -- RECORD_TEST_END
+
+You'll notice that the syntax to declare records is very similar to the syntax used to declare templates. That's no accident because a template is really just a special record. When you write ``template Token with``, one of the things that happens in the background is that this becomes a ``data Token = Token with``.
+
+In the ``assert`` statements above, we always compared values of in-built types. If you wrote ``assert (my_record == my_record)`` in the scenario, you may be surprised to get an error message ``No instance for (Eq MyRecord) arising from a use of ‘==’``. Equality in DAML is always value equality and we haven't written a function to check value equality for ``MyRecord`` values. But don't worry, you don't have to implement this rather obvious function yourself. The compiler is smart enough to do it for you.
+
+.. literalinclude:: daml/3_Data/Record.daml
+  :language: daml
+  :start-after: -- EQ_TEST_BEGIN
+  :end-before: -- EQ_TEST_END
+
+``Eq`` is a so-called type-class, which can be thought of like an interface in other languages. There are some other type-classes that can be derived automatically. Most prominently ``Show`` to get access to the function ``show``, equivalent to ``toString`` in many languages, and ``Ord``, which gives access to comparison operators ``<``, ``>``, ``<=``, ``>=``.
+
+It's a good idea to always derive ``Eq`` and ``Show`` using ``deriving (Eq, Show)``. The record types created using ``template T with`` do this automatically.
+
+Records can give the data on ``CashBalance`` a bit more structure:
+
+.. literalinclude:: daml/3_Data/Record.daml
+  :language: daml
+  :start-after: -- CASH_BALANCE_BEGIN
+  :end-before: -- CASH_BALANCE_END
+
+If you look at the resulting scenario view, you'll see that this still gives rise to one table. The records are expanded out into columns using dot notation.
+
+Variants
+~~~~~~~~
+
+Suppose now that you also wanted to keep track of cash in hand. Cash in hand doesn't have a bank, but you can't just leave ``bank`` empty. DAML doesn't have an equivalent to ``null``. Variants can express that cash can either be in hand or at a bank.
+
+.. literalinclude:: daml/3_Data/Variants.daml
+  :language: daml
+  :start-after: -- CASH_BALANCE_BEGIN
+  :end-before: -- CASH_BALANCE_END
+
+The way to read the declaration of ``Location`` is "*A Location either has value* ``InHand`` *OR has a value* ``InAccount a`` *where* ``a`` *is of type Account*". This is quite an explicit way to say that there may or may not be an ``Account`` associated with a ``CashBalance`` and gives both cases suggestive names. The generic way would be to use the built-in ``Optional`` type. The ``None`` value of type ``Optional a`` is the closest DAML has to a ``null`` value.
```

Comment:
```suggestion
The way to read the declaration of ``Location`` is "*A Location either has value* ``InHand`` *OR has a value* ``InAccount a`` *where* ``a`` *is of type Account*". This is quite an explicit way to say that there may or may not be an ``Account`` associated with a ``CashBalance`` and gives both cases suggestive names. Another option would be to use the built-in ``Optional`` type. The ``None`` value of type ``Optional a`` is the closest DAML has to a ``null`` value.
```
"The generic way" does not mean anything to me as a reader at this point. I don't see why using Optional is "more generic".